### PR TITLE
DAOS-12961 cart: return error if crt_provider_get_ctx_idx() failed

### PR DESCRIPTION
**Pull request by wangshilong**
Fix to return error for crt_provider_get_ctx_idx(), so we don't need
call crt_provider_put_ctx_idx() if it failed to avoid assertion in crt_finalize().

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by wiliamhuang — APPROVED** [review]

File: src/cart/crt_context.c [hunk 1]
```diff
@@ -229,6 +229,8 @@ crt_context_provider_create(crt_context_t *crt_ctx, crt_provider_t provider, boo
 	if (ctx_idx < 0 || ctx_idx >= max_ctx_num) {
 		D_WARN("Provider: %d; Context limit (%d) reached\n",
 		       provider, max_ctx_num);
+		if (ctx_idx >=0)
+			crt_provider_put_ctx_idx(primary, provider, ctx_idx);
```
Comment (frostedcmos) [inline src/cart/crt_context.c:233]: This does not look correct to me. if we hit this case, means ctx_idx was >= max_ctx_num, in other words it was invalid.

We dont need to put invalid context back, and doing so today in fact might cause a crash as we access array at the invalid index.

603 void
 604 crt_provider_put_ctx_idx(bool primary, int provider, int idx)
 605 {
 606         struct crt_prov_gdata *prov_data = crt_get_prov_gdata(primary, provider);
 607
 608         if (prov_data->cpg_used_idx[idx] == false) {

**Review by frostedcmos — CHANGES_REQUESTED** [review]

**Comment by wiliamhuang** [comment]
@frostedcmos https://github.com/daos-stack/daos/blob/master/src/cart/crt_init.c#L217-L223
It seems that max_num_ctx (or prov_data->cpg_ctx_max_num) could be smaller than CRT_SRV_CONTEXT_NUM. If this is true, the return value of crt_provider_get_ctx_idx() could be larger than max_ctx_num.

**Comment by frostedcmos** [comment]
> @frostedcmos https://github.com/daos-stack/daos/blob/master/src/cart/crt_init.c#L217-L223 It seems that max_num_ctx (or prov_data->cpg_ctx_max_num) could be smaller than CRT_SRV_CONTEXT_NUM. If this is true, the return value of crt_provider_get_ctx_idx() could be larger than max_ctx_num.

Good point there, however this then tells me we should fix crt_provider_get_ctx_idx() instead and return -1 for any invalid case.  The overall idea is that if 'get' returns an error, then put should not be required.

So for get(), i would say it should only loop up to prov_data->cpg_ctx_max_num instead of going through CRT_SRV_CONTEXT_NUM.

**Comment by wiliamhuang** [comment]
> So for get(), i would say it should only loop up to prov_data->cpg_ctx_max_num instead of going through CRT_SRV_CONTEXT_NUM.

I agree. Thank you!

**Comment by wangshilong** [comment]
Thanks @frostedcmos and @wiliamhuang , sorry i missed to force-update the PR,  but it might be ok since change is trival.

**Review by frostedcmos — APPROVED** [review]

**Review by wiliamhuang — APPROVED** [review]

**Pull request merged**